### PR TITLE
Make minBufferBindingSize optional

### PR DESCRIPTION
**Pull request by litherum**
When the author opt-in to using minBufferBindingSize, this enables two checks:
1. The appropriate buffer in any bind groups created with this bind group layout is larger than minBufferBindingSize
2. minBufferBindingSize is larger than is necessary given the shader

If the author opts in to these requirements, we can do bounds checking at bind group creation / shader compilation time. However, if the author does not opt in to these requirements, we must do bounds checking at draw call time.

Rather than using a sentinel 0 for the second check, we should have the author explicit enable them via out-of-band messaging. This doesn't matter for the first check (since any buffer length is greater-than-or-equal-to 0), but it does matter for the second check.

This patch doesn't change any defaults; it just makes makes the opt-in to earlier bounds checking more explicit.

***
<a href="https://pr-preview.s3.amazonaws.com/litherum/gpuweb/pull/874.html" title="Last updated on Jul 20, 2020, 9:49 PM UTC (0d2a054)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/874/ec5cda7...litherum:0d2a054.html" title="Last updated on Jul 20, 2020, 9:49 PM UTC (0d2a054)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -1982,7 +1982,7 @@ dictionary GPUBindGroupLayoutEntry {
     boolean hasDynamicOffset = false;
 
     // Used for uniform buffer and storage buffer bindings.
-    GPUSize64 minBufferBindingSize = 0;
+    GPUSize64? minBufferBindingSize = null;
```
Comment (Kangz) [inline spec/index.bs:1985]: nit: this could just be `GPUSize64 minBufferBindingSize`.

**Comment by Kangz** [comment]
I have slight concerns about using `minBufferBindingSize` not being present as the signal it shouldn't be used because WASM doesn't have optional dictionary members. So far the spec doesn't have optional numbers that cannot be replaced just by 0, except `GPUBufferBinding.size` which is already slightly unfortunate to use from WASM and requires a `UINT64_MAX` sentinel value or a similar mechanism.

**Comment by kvark** [comment]
@Kangz I share the concern about WASM :) Also:

> except GPUBufferBinding.size which is already slightly unfortunate to use from WASM and requires a UINT64_MAX sentinel value or a similar mechanism.

Note that with `minBufferBindingSize` agreed in CG, it does no longer make sense to support zero-sized buffer bindings: they have to be larger or equal to the non-zero `minBufferBindingSize` anyway. So making `size=0` to mean "rest of the buffer" is an option on the able for WASM.

**Comment by kainino0x** [comment]
> Rather than using a sentinel 0 for the second check, we should have the author explicit enable them via out-of-band messaging. This doesn't matter for the first check (since any buffer length is greater-than-or-equal-to 0), but it does matter for the second check.

I argued against this already somewhere. It's one thing to use 0 as a sentinel value. But in this case we really do literally mean "the minimum buffer binding size is 0". So in terms of the local semantics, it matches the name and I think we should keep that, despite the fact that no-minimum-buffer-binding-size does become special later.

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:1985]:
```suggestion
    GPUSize64 minBufferBindingSize;
```

File: spec/index.bs [hunk 2]
```diff
@@ -2142,7 +2142,7 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
             {{GPUBindingType/"readonly-storage-buffer"}},
             ensure
             |bindingDescriptor|.{{GPUBindGroupLayoutEntry/hasDynamicOffset}} is `false`,
-            and |bindingDescriptor|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is zero.
+            and |bindingDescriptor|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is null.
```
Comment (kainino0x) [inline spec/index.bs:2145]:
```suggestion
            and |bindingDescriptor|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is undefined.
```

**Comment by litherum** [comment]
Right now, the spec has two situations:
1. Passing a value that is greater than 0 as `minBufferBindingSize`, which enables ahead-of-time bounds checking
2. Either passing 0 or not passing anything as `minBufferBindingSize`, which enables draw-call bounds checking

This patch moves the "passing 0" into the first situation.

The reason for this change is that it's totally reasonable for a shader to have no resources bound other than a big array. In this situation, you don't want draw-call bounds checking, but the value to check against is 0.

The switch between "do ahead-of-time bounds checking" and "do draw-call bounds checking" should be an explicit switch, not implied by a sentinel 0.

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:1985]: Marking unresolved as this doesn't seem to have actually been done

File: spec/index.bs [hunk 3]
```diff
@@ -2781,7 +2782,7 @@ A {{GPUProgrammableStageDescriptor}} describes the entry point in the user-provi
             the |binding| has to be a read-only storage buffer.
         1. If |entry|.{{GPUBindGroupLayoutEntry/type}} is {{GPUBindingType/"sampled-texture"}}, {{GPUBindingType/"readonly-storage-texture"}}, or {{GPUBindingType/"writeonly-storage-texture"}},
             the shader view dimension of the texture has to match |entry|.{{GPUBindGroupLayoutEntry/viewDimension}}.
-        1. If |entry|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is not zero:
+        1. If |entry|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is not null:
```
Comment (kainino0x) [inline spec/index.bs:2785]:
```suggestion
        1. If |entry|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}} is not undefined:
```
oops, I missed this

File: spec/index.bs [hunk 4]
```diff
@@ -2398,7 +2398,8 @@ A {{GPUBindGroup}} object has the following internal slots:
                 with {{GPUBufferUsage/STORAGE}} flag.
             1. The bound part designated by |bufferBinding|.{{GPUBufferBinding/offset}} and
                 |bufferBinding|.{{GPUBufferBinding/size}} must reside inside the buffer.
-            1. The effective binding size, that is either explict in |bufferBinding|.{{GPUBufferBinding/size}}
+            1. If |layoutBinding|.{{GPUBindGroupLayoutEntry/minBufferBindingSize|GPUBindGroupLayoutEntry.minBufferBindingSize}} is not null,
```
Comment (kainino0x) [inline spec/index.bs:2401]:
```suggestion
            1. If |layoutBinding|.{{GPUBindGroupLayoutEntry/minBufferBindingSize|GPUBindGroupLayoutEntry.minBufferBindingSize}} is not undefined,
```

File: spec/index.bs [hunk 5]
```diff
@@ -2398,7 +2398,8 @@ A {{GPUBindGroup}} object has the following internal slots:
...
+                The effective binding size, that is either explict in |bufferBinding|.{{GPUBufferBinding/size}}
                 or derived from |bufferBinding|.{{GPUBufferBinding/offset}} and the full size of the buffer,
                 is greater than or equal to |layoutBinding|.{{GPUBindGroupLayoutEntry/minBufferBindingSize|GPUBindGroupLayoutEntry.minBufferBindingSize}}.
```
Comment (kainino0x) [inline spec/index.bs:2404]:
```suggestion
            1. If |layoutBinding|.{{GPUBindGroupLayoutEntry/minBufferBindingSize|GPUBindGroupLayoutEntry.minBufferBindingSize}} is not undefined:
                - The effective binding size, that is either explict in |bufferBinding|.{{GPUBufferBinding/size}}
                    or derived from |bufferBinding|.{{GPUBufferBinding/offset}} and the full size of the buffer,
                    is greater than or equal to |layoutBinding|.{{GPUBindGroupLayoutEntry/minBufferBindingSize|GPUBindGroupLayoutEntry.minBufferBindingSize}}.
```

**Comment by kvark** [comment]
> The reason for this change is that it's totally reasonable for a shader to have no resources bound other than a big array. In this situation, you don't want draw-call bounds checking, but the value to check against is 0.

On the call we clarified that this is impossible today, as we agreed to require the buffer binding to include at least one element of an unsized array. So there is *never* a case where you'd want to specify 0 there while not wanting draw-call bounds checks.

I assumed that you were going to revise this proposal with that aspect clarified, but it appears to be unchanged.

**Comment by kainino0x** [comment]
What needs clarification? That 0 is not a valid value?

**Comment by kvark** [comment]
Nothing needs clarification, it was provided on the call and that's what I wrote in the comment.
It just seems to me that clarification removes one of the concerns that this PR is trying to solve, and therefore I expected the PR to be reconsidered.

**Comment by kainino0x** [comment]
Right, I see now that Myles's hasn't come back to this issue since the meeting. Doesn't change that I'm still fine with this if Myles wants to push it forward though.

0 would never be a valid minBufferBindingSize because it would fail to validate against any shader during pipeline creation. So if we accept this, we might want to add validation in BGL creation saying minBufferBindingSize cannot be 0.

**Comment by litherum** [comment]
Presumably the set of resources attached to the shader (via bindgroups) is allowed to be a superset of the resources that the shader actually uses. For the resources which are bound but not used, 0 seems like a reasonable value.

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs]: I committed this suggestion

**Pull request merged**